File: day-04/solution.ts
import { readLines } from "../commons/io.ts";

function main() {
  const lines = readLines("day-04/input.txt");
  let duplicateRanges = 0;
  let overlapRanges = 0;

  class Range {
    public min: number;
    public max: number;
    constructor(range: string) {
      const [min, max] = range.split("-").map((n) => parseInt(n));
      this.min = min;
      this.max = max;
    }
    isContaining(other: Range): boolean {
      return this.min <= other.min && this.max >= other.max;
    }
    isOverlapping(other: Range): boolean {
      let overlap = false;
      if (this.min <= other.min && this.max >= other.min) overlap = true;
      if (this.min <= other.max && this.max >= other.max) overlap = true;
      return overlap;
    }
  }

  lines.forEach((line) => {
    const [range1, range2] = line.split(",").map((r) => new Range(r));
    if (range1.isContaining(range2) || range2.isContaining(range1)) {
      duplicateRanges++;
    }

    if (range1.isOverlapping(range2) || range2.isOverlapping(range1)) {
      overlapRanges++;
    }
  });

  console.log(`Duplicate ranges: ${duplicateRanges}`);
  console.log(`Overlapping ranges: ${overlapRanges}`);
}

main();

File: day-04/solution.py
import sys
import os

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

from commons.io import readLines


class Range:
  def __init__(self, range_str: str):
    self.min, self.max = map(int, range_str.split("-"))

  def is_containing(self, other: "Range") -> bool:
    return self.min <= other.min and self.max >= other.max

  def is_overlapping(self, other: "Range") -> bool:
    overlap = False
    if self.min <= other.min and self.max >= other.min:
      overlap = True
    if self.min <= other.max and self.max >= other.max:
      overlap = True
    return overlap

def main():
  lines = readLines("day-04/input.txt")
  duplicate_lines = 0
  overlap_lines = 0

  for line in lines:
    range1, range2 = map(lambda r: Range(r), line.split(","))
    if range1.is_containing(range2) or range2.is_containing(range1):
      duplicate_lines += 1

    if range1.is_overlapping(range2) or range2.is_overlapping(range1):
      overlap_lines += 1

  print(f"Duplicate lines: {duplicate_lines}")
  print(f"Overlap lines: {overlap_lines}")


main()
